### PR TITLE
summary failedtests;test=develop

### PR DESCRIPTION
**Pull request by lelelelelez**
### PR types
 Others
### PR changes

Others
### Describe

执行失败单测整合到一起展示
测试结果如下：
coverage:https://xly.bce.baidu.com/paddlepaddle/paddle/ipipe/#/detail/1248469/job/1776316
py35: https://xly.bce.baidu.com/paddlepaddle/paddle/ipipe/#/detail/1248068/job/1775734
image

### CODE REVIEW

**Comment by luotao1** [comment]
目前会打出两段，且前面有空格
```
....
2020-07-06 13:59:28 failuretestxxx: 	903 - test_listen_and_serv_op (Failed)
2020-07-06 13:59:28 summary_failtestxxx: 
2020-07-06 13:59:28             	633 - test_mul_op (Failed)
2020-07-06 13:59:28             	677 - test_parallel_executor_seresnext_with_fuse_all_reduce_gpu (Failed)
2020-07-06 13:59:28             	903 - test_listen_and_serv_op (Failed)
2020-07-06 13:59:28         ========================================
2020-07-06 13:59:28         Summary Failed Tests ...
2020-07-06 13:59:28         ========================================
2020-07-06 13:59:28 
2020-07-06 13:59:28             	633 - test_mul_op (Failed)
2020-07-06 13:59:28             	677 - test_parallel_executor_seresnext_with_fuse_all_reduce_gpu (Failed)
2020-07-06 13:59:28             	903 - test_listen_and_serv_op (Failed)
```
能否改成
```
...
2020-07-06 13:59:28 ========================================
2020-07-06 13:59:28 Summary Failed Tests ...
2020-07-06 13:59:28 ========================================
2020-07-06 13:59:28     633 - test_mul_op (Failed)
2020-07-06 13:59:28     677 - test_parallel_executor_seresnext_with_fuse_all_reduce_gpu (Failed)
2020-07-06 13:59:28     903 - test_listen_and_serv_op (Failed)
```

File: paddle/scripts/paddle_build.sh [hunk 1]
```diff
@@ -744,6 +744,25 @@ EOF
...
 
+summary_failtest=''
+tmpdir=`mktemp -d`
+
+function gather_failtests() {
+    for file in `ls $tmpdir`; do
+        exit_code=0
+        grep -q 'The following tests FAILED:' $tmpdir/$file||exit_code=$?
+        if [ $exit_code -ne 0 ]; then
+            #echo "$exit_code != 0"
```
Comment (luotao1) [inline paddle/scripts/paddle_build.sh]: 去掉注释的755和758行

File: paddle/scripts/paddle_build.sh [hunk 2]
```diff
@@ -744,6 +744,25 @@ EOF
     fi
 }
 
+summary_failtest=''
```
Comment (luotao1) [inline paddle/scripts/paddle_build.sh]: 变量名可以更加规范一点：
summary_failtest-> failed_test_lists?
tmpdir-> tmp_dir?

File: paddle/scripts/paddle_build.sh [hunk 3]
```diff
@@ -744,6 +744,25 @@ EOF
...
+tmpdir=`mktemp -d`
+
+function gather_failtests() {
```
Comment (luotao1) [inline paddle/scripts/paddle_build.sh]: gather_failtests-》collect_failed_tests?

File: paddle/scripts/paddle_build.sh [hunk 4]
```diff
@@ -779,21 +798,22 @@ function card_test() {
...
             fi
         done
+        tmpfile_rand=`date +%s%N`
+        tmpfile=$tmpdir/$tmpfile_rand
         if [ ${TESTING_DEBUG_MODE:-OFF} == "ON" ] ; then
             if [[ $cardnumber == $CUDA_DEVICE_COUNT ]]; then
-                ctest -I $i,,$NUM_PROC -R "($testcases)" -V &
-            else
-                env CUDA_VISIBLE_DEVICES=$cuda_list ctest -I $i,,$NUM_PROC -R "($testcases)" -V &
+                (ctest -I $i,,$NUM_PROC -R "($testcases)" -V | tee $tmpfile; test ${PIPESTATUS[0]} -eq 0) &
```
Comment (luotao1) [inline paddle/scripts/paddle_build.sh:802]: - `tee $tmpfile; test ${PIPESTATUS[0]} -eq 0`可以封装成一个函数么？
- 其中`test ${PIPESTATUS[0]} -eq 0`是用来做什么的？

File: paddle/scripts/paddle_build.sh (hunk 1 above)
Comment (lelelelelez) [reply paddle/scripts/paddle_build.sh]: done

File: paddle/scripts/paddle_build.sh (hunk 2 above)
Comment (lelelelelez) [reply paddle/scripts/paddle_build.sh]: done

File: paddle/scripts/paddle_build.sh (hunk 3 above)
Comment (lelelelelez) [reply paddle/scripts/paddle_build.sh]: done

File: paddle/scripts/paddle_build.sh (hunk 4 above)
Comment (lelelelelez) [reply paddle/scripts/paddle_build.sh:802]: 1. `ctest -I $i,,$NUM_PROC -R "($testcases)" -V | tee $tmpfile; test ${PIPESTATUS[0]} -eq 0`是为了将ctest的结果不仅打在终端，而且可以重定下到tmpfile文件中，为了后期分析tmpfile的失败case，一气呵成的，不建议封装成函数。
2. card_test()中有个`trap 'caught_error' CHLD` 目的是为了捕捉ctest的子进程中失败case，而我们为了要将ctest的结果打到tmpfile中，`ctest -I $i,,$NUM_PROC -R "($testcases)" -V | tee $tmpfile;`是一个整体，它的返回码会一直是0，`trap 'caught_error' CHLD`将捕捉不到失败case，因此使用`PIPESTATUS`来获取管道中所有命令的返回码。

**Comment by lelelelelez** [comment]
> 目前会打出两段，且前面有空格
> 
> ```
> ....
> 2020-07-06 13:59:28 failuretestxxx: 	903 - test_listen_and_serv_op (Failed)
> 2020-07-06 13:59:28 summary_failtestxxx: 
> 2020-07-06 13:59:28             	633 - test_mul_op (Failed)
> 2020-07-06 13:59:28             	677 - test_parallel_executor_seresnext_with_fuse_all_reduce_gpu (Failed)
> 2020-07-06 13:59:28             	903 - test_listen_and_serv_op (Failed)
> 2020-07-06 13:59:28         ========================================
> 2020-07-06 13:59:28         Summary Failed Tests ...
> 2020-07-06 13:59:28         ========================================
> 2020-07-06 13:59:28 
> 2020-07-06 13:59:28             	633 - test_mul_op (Failed)
> 2020-07-06 13:59:28             	677 - test_parallel_executor_seresnext_with_fuse_all_reduce_gpu (Failed)
> 2020-07-06 13:59:28             	903 - test_listen_and_serv_op (Failed)
> ```
> 
> 能否改成
> 
> ```
> ...
> 2020-07-06 13:59:28 ========================================
> 2020-07-06 13:59:28 Summary Failed Tests ...
> 2020-07-06 13:59:28 ========================================
> 2020-07-06 13:59:28     633 - test_mul_op (Failed)
> 2020-07-06 13:59:28     677 - test_parallel_executor_seresnext_with_fuse_all_reduce_gpu (Failed)
> 2020-07-06 13:59:28     903 - test_listen_and_serv_op (Failed)
> ```

done

**Review by luotao1 — APPROVED** [review]
LGTM

**Pull request merged**